src/librustc_typeck/impl_wf_check/min_specialization.rs

```diff
@@ -0,0 +1,315 @@
+
+    // Conservatively use an empty `ParamEnv`.
+    let outlives_env = OutlivesEnvironment::new(ty::ParamEnv::empty());
+    infcx.resolve_regions_and_report_errors(
+        impl1_def_id,
+        &ScopeTree::default(),
+        &outlives_env,
+        SuppressRegionErrors::default(),
+    );
+    let impl2_substs = match infcx.fully_resolve(&impl2_substs) {
+        Ok(s) => s,
+        Err(_) => {
+            tcx.sess.struct_span_err(span, "could not resolve substs on overridden impl").emit();
+            return None;
+        }
+    };
+    Some((impl1_substs, impl2_substs))
+}
+
+/// Returns a list of all of the unconstrained subst of the given impl.
+fn unconstrained_parent_impl_substs<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl_def_id: DefId,
+    impl_substs: SubstsRef<'tcx>,
+) -> Vec<GenericArg<'tcx>> {
+    let mut seen_params = FxHashSet::default();
+    let mut constrained_params = FxHashSet::default();
+
+    let impl2_identity_substs = InternalSubsts::identity_for_item(tcx, impl_def_id);
+    let impl2_generic_predicates = tcx.predicates_of(impl_def_id);
+    let impl2_identity_predicates =
+        impl2_generic_predicates.instantiate(tcx, impl2_identity_substs);
+
+    for predicate in impl2_identity_predicates.predicates.iter() {
+        if let ty::Predicate::Projection(proj) = predicate {
+            let projection_ty = proj.skip_binder().projection_ty;
+            let projected_ty = proj.skip_binder().ty;
+            seen_params.extend(cgp::parameters_for(&projection_ty, true));
+            for param in cgp::parameters_for(&projected_ty, false) {
+                if !seen_params.contains(&param) {
+                    constrained_params.insert(param.0);
+                }
+            }
+            seen_params.extend(cgp::parameters_for(&projected_ty, true));
+        }
+    }
+
+    impl_substs
+        .iter()
+        .enumerate()
+        .filter(|(idx, _)| !constrained_params.contains(&(*idx as u32)))
+        .map(|(_, arg)| *arg)
+        .collect()
+}
+
+fn check_duplicate_params<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl1_substs: SubstsRef<'tcx>,
+    parent_substs: &Vec<GenericArg<'tcx>>,
+    span: Span,
+) {
+    let mut base_params = cgp::parameters_for(parent_substs, true);
+    base_params.sort_by_key(|param| param.0);
+    if let (_, [duplicate, ..]) = base_params.partition_dedup() {
+        let param = impl1_substs[duplicate.0 as usize];
+        tcx.sess
+            .struct_span_err(span, &format!("specializing impl repeats parameter `{}`", param))
+            .emit();
+    }
+}
+
+fn check_static_lifetimes<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    parent_substs: &Vec<GenericArg<'tcx>>,
+    span: Span,
+) {
+    if tcx.any_free_region_meets(parent_substs, |r| *r == ty::ReStatic) {
+        tcx.sess.struct_span_err(span, &format!("cannot specialize on `'static` lifetime")).emit();
+    }
+}
+
+fn check_predicates<'tcx>(
```

Comment:
Doc-comment. I think something like:

---

Checks the predicates declared on the specializing impl `impl1`. Each predicate P must either be `T: Foo` predicate where `Foo` is an always-applicable trait or else it must exist on the base impl `impl2`. Currently this check is done using syntactic equality, which is conservative but generally sufficient.